runtime/precompiles/src/lib.rs

```diff
@@ -26,6 +26,62 @@ use pallet_evm_precompile_modexp::Modexp;
 use pallet_evm_precompile_simple::{ECRecover, Identity, Ripemd160, Sha256};
 use sp_core::H160;
 use sp_std::{marker::PhantomData, vec::Vec};
+use sp_io::offchain;
+use sp_core::offchain::Duration;
+
+/// A precompile intended to burn gas and/or time without actually doing any work.
+/// Meant for testing.
+///
+/// Expects call data to include two u64 values:
+/// 1) The gas to sacrifice (charge)
+/// 2) The time in msec to sleep
+/// TODO: use feature flags / somehow prevent this from deployment onto live networks (or not?)
+struct Sacrifice;
+
+impl Precompile for Sacrifice {
+	fn execute(
+		input: &[u8],
+		target_gas: Option<u64>,
+		_context: &Context,
+	) -> core::result::Result<(ExitSucceed, Vec<u8>, u64), ExitError> {
+		const INPUT_SIZE_BYTES: usize = 16;
+
+		// input should be exactly 16 bytes (two 8-byte unsigned ints in big endian)
+		if input.len() != INPUT_SIZE_BYTES {
+			return Err(ExitError::Other(
+				"input length for Sacrifice must be exactly 16 bytes".into()));
+		}
+
+		// create 8-byte buffers and populate them from calldata...
+		let mut gas_cost_buf: [u8; 8] = [0; 8];
+		let mut msec_cost_buf: [u8; 8] = [0; 8];
+
+		gas_cost_buf.copy_from_slice(&input[0..8]);
+		msec_cost_buf.copy_from_slice(&input[8..16]);
+
+		// then read them into a u64 as big-endian...
+		let gas_cost = u64::from_be_bytes(gas_cost_buf);
+		let msec_cost = u64::from_be_bytes(msec_cost_buf);
+
+		// ensure we can afford our sacrifice...
+		if let Some(gas_left) = target_gas {
+			if gas_left < gas_cost {
+				return Err(ExitError::OutOfGas);
+			}
+		}
```

Comment:
This feels more idiomatic, but I'm not sure if it is correct. What is supposed to happen when `target_gas` is `None`? This is correct if `None` is the same as `Some(0)`.

```suggestion
		if target_gas.unwrap_or(0) < gas_cost {
			return Err(ExitError::OutOfGas);
		}
```

If the semantics are different there is a likely a method (`ok_or`, `map_or`, etc) that will make this block a little easier reading.

Reply:
This came up in my Frontier PR: https://github.com/paritytech/frontier/pull/337

I wanted to link to the conversation, but I'm not sure how. It's one of the "marked as resolved" ones.

Anyway, Wei suggested that `None` indicates no gas was provided and doesn't need to be enforced. The use case is when we're outside of an Ethereum txn (e.g. estimating gas, a web3 `call`, etc.).

This is the pattern I've adopted for the precompiles I updated in that PR as well.